File: app/src/main/java/com/veganbeauty/app/data/repository/OrderStatusNotifier.java
package com.veganbeauty.app.data.repository;

import android.content.Context;
import android.util.Log;

import com.veganbeauty.app.data.local.ProfileSession;
import com.veganbeauty.app.data.local.dao.OrderDao;
import com.veganbeauty.app.data.local.entities.NotificationItem;
import com.veganbeauty.app.data.local.entities.OrderEntity;
import com.veganbeauty.app.features.account.notification.NotificationDateHelper;
import com.veganbeauty.app.features.account.notification.NotificationPushHelper;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public final class OrderStatusNotifier {

    private static final String TAG = "OrderStatusNotifier";
    private static final ExecutorService executor = Executors.newSingleThreadExecutor();

    private OrderStatusNotifier() {
    }

    public static void updateStatus(OrderDao orderDao, Context context, String orderId, String newStatus) {
        executor.execute(() -> {
            try {
                OrderEntity existing = orderDao.getOrderByIdSync(orderId);
                String previousStatus = existing != null ? existing.getStatus() : null;
                int updated = orderDao.updateOrderStatusSync(orderId, newStatus);
                if (updated <= 0) {
                    Log.w(TAG, "updateStatus: no order found for id=" + orderId);
                    return;
                }
                OrderEntity order = orderDao.getOrderByIdSync(orderId);
                if (order != null) {
                    ensureStatusNotification(context, order);
                }
            } catch (Exception e) {
                Log.e(TAG, "updateStatus failed for orderId=" + orderId, e);
            }
        });
    }

    public static void simulateOnly(OrderDao orderDao, Context context, String orderId, String newStatus) {
        try {
            OrderEntity order = orderDao.getOrderByIdSync(orderId);
            if (order == null) {
                return;
            }
            order.setStatus(newStatus);
            ensureStatusNotification(context, order);
        } catch (Exception e) {
            Log.e(TAG, "simulateOnly failed for orderId=" + orderId, e);
        }
    }

    /**
     * Sends order notifications based on the current status and what was already notified.
     * Safe to call repeatedly — skips statuses the user was already notified about.
     */
    public static void ensureStatusNotification(Context context, OrderEntity order) {
        if (order == null || order.getId() == null || order.getId().trim().isEmpty()) {
            return;
        }
        String newStatus = order.getStatus();
        if (newStatus == null || newStatus.trim().isEmpty()) {
            return;
        }

        boolean isGuest = order.isGuest()
                || order.getUserId() == null
                || order.getUserId().trim().isEmpty();
        if (isGuest) {
            simulateOutboundNotification(order, newStatus);
            return;
        }

        if (context == null) {
            Log.w(TAG, "Missing context for order notification " + order.getId());
            return;
        }

        Context appContext = context.getApplicationContext();
        if (!OrderNotificationTracker.needsNotification(appContext, order.getId(), newStatus)) {
            return;
        }

        if (!ProfileSession.isOrderStatusEnabled(appContext)) {
            return;
        }

        String normalized = OrderNotificationTracker.normalize(newStatus);
        if ("pending".equals(normalized)
                && OrderNotificationTracker.getLastNotifiedStatus(appContext, order.getId()) == null) {
            notifyOrderPlaced(appContext, order);
            return;
        }

        String previousNotified = OrderNotificationTracker.getLastNotifiedStatus(appContext, order.getId());
        notifyIfStatusChanged(appContext, order, previousNotified, newStatus);
    }

    public static void notifyOrderPlaced(Context context, OrderEntity order) {
        if (order == null || order.getId() == null || order.getId().trim().isEmpty()) {
            return;
        }

        boolean isGuest = order.isGuest()
                || order.getUserId() == null
                || order.getUserId().trim().isEmpty();
        if (isGuest) {
            simulateOutboundNotification(order, order.getStatus() != null ? order.getStatus() : "Chờ xác nhận");
            return;
        }
        if (context == null) {
            Log.w(TAG, "Missing context for order placed notification " + order.getId());
            return;
        }

        Context appContext = context.getApplicationContext();
        if (!ProfileSession.isOrderStatusEnabled(appContext)) {
            return;
        }

        String orderId = order.getId().trim();
        String status = order.getStatus() != null ? order.getStatus() : "Chờ xác nhận";
        if (!OrderNotificationTracker.needsNotification(appContext, orderId, status)) {
            return;
        }

        String notificationId = "order_placed_" + orderId;
        NotificationRepository repository = NotificationRepository.getInstance(appContext);
        for (NotificationItem existing : repository.notifications.getValue()) {
            if (notificationId.equals(existing.getId())) {
                OrderNotificationTracker.markNotified(appContext, orderId, status);
                return;
            }
        }

        String title = "Đơn hàng đặt thành công! \uD83D\uDED2";
        String content = buildMessageBody(orderId, "đã được ghi nhận và đang chờ xác nhận");
        dispatchInApp(
                appContext,
                orderId,
                notificationId,
                title,
                content,
                resolveOrderPlacementTime(order),
                status
        );
    }

    public static void notifyIfStatusChanged(
            Context context,
            OrderEntity order,
            String previousStatus,
            String newStatus
    ) {
        if (order == null || newStatus == null || newStatus.trim().isEmpty()) {
            return;
        }

        boolean isGuest = order.isGuest()
                || order.getUserId() == null
                || order.getUserId().trim().isEmpty();
        if (isGuest) {
            simulateOutboundNotification(order, newStatus);
            return;
        }

        if (context == null) {
            Log.w(TAG, "Missing context for in-app order notification " + order.getId());
            return;
        }

        Context appContext = context.getApplicationContext();
        if (!ProfileSession.isOrderStatusEnabled(appContext)) {
            return;
        }
        if (!OrderNotificationTracker.needsNotification(appContext, order.getId(), newStatus)) {
            return;
        }

        NotificationContent notificationContent = resolveNotificationContent(order.getId(), newStatus);
        String notificationId = "order_status_" + order.getId() + "_" + System.currentTimeMillis();
        dispatchInApp(
                appContext,
                order.getId(),
                notificationId,
                notificationContent.title,
                notificationContent.content,
                currentEventTime(),
                newStatus
        );
    }

    private static void dispatchInApp(
            Context appContext,
            String orderId,
            String notificationId,
            String title,
            String content,
            String eventTime,
            String trackedStatus
    ) {
        NotificationItem item = new NotificationItem(
                notificationId,
                title,
                content,
                eventTime,
                "Đơn hàng",
                null,
                null,
                "CHI TIẾT",
                false,
                "Hôm nay",
                "ic_bell",
                "order",
                orderId,
                null
        );

        NotificationRepository.getInstance(appContext).addNotification(item);
        NotificationPushHelper.sendPushNotification(
                appContext,
                notificationId,
                title,
                content,
                "order",
                null,
                orderId,
                null
        );
        OrderNotificationTracker.markNotified(appContext, orderId, trackedStatus);
        Log.i(TAG, "Order notification dispatched for " + orderId + ": " + title);
    }

    public static void backfillPendingOrderNotifications(
            Context context,
            OrderDao orderDao,
            String userId
    ) {
        if (context == null || orderDao == null || userId == null || userId.trim().isEmpty()) {
            return;
        }
        try {
            List<OrderEntity> pendingOrders = orderDao.getPendingOrdersForUserSync(userId.trim());
            if (pendingOrders == null) {
                return;
            }
            for (OrderEntity order : pendingOrders) {
                notifyOrderPlaced(context, order);
            }
        } catch (Exception e) {
            Log.e(TAG, "backfillPendingOrderNotifications failed", e);
        }
    }

    /**
     * Re-checks local orders against last notified status — no extra Firestore reads.
     */
    public static void backfillMissedOrderStatusNotifications(
            Context context,
            OrderDao orderDao,
            String userId,
            String phone
    ) {
        if (context == null || orderDao == null) {
            return;
        }
        String safeUserId = userId != null ? userId.trim() : "";
        String safePhone = phone != null ? phone.trim() : "";
        if (safeUserId.isEmpty() && safePhone.isEmpty()) {
            return;
        }
        try {
            List<OrderEntity> orders = orderDao.getOrdersForBuyerIdentitySync(safeUserId, safePhone);
            if (orders == null || orders.isEmpty()) {
                return;
            }
            for (OrderEntity order : orders) {
                ensureStatusNotification(context, order);
            }
        } catch (Exception e) {
            Log.e(TAG, "backfillMissedOrderStatusNotifications failed", e);
        }
    }

    private static NotificationContent resolveNotificationContent(String orderId, String status) {
        switch (OrderNotificationTracker.normalize(status)) {
            case "processing":
                return new NotificationContent(
                        "Đơn hàng đã được xác nhận",
                        buildMessageBody(orderId, "đã được xác nhận và đang được chuẩn bị")
                );
            case "delivering":
                return new NotificationContent(
                        "Đơn hàng đang được giao",
                        buildMessageBody(orderId, "đang được giao đến bạn")
                );
            case "delivered":
                return new NotificationContent(
                        "Đơn hàng đã giao thành công",
                        buildMessageBody(orderId, "đã được giao thành công")
                );
            case "cancelled":
                return new NotificationContent(
                        "Đơn hàng đã bị huỷ",
                        buildMessageBody(orderId, "đã được huỷ theo yêu cầu")
                );
            default:
                return new NotificationContent(
                        "Cập nhật đơn hàng",
                        buildMessageBody(orderId, "đã được cập nhật trạng thái: " + status)
                );
        }
    }

    private static void simulateOutboundNotification(OrderEntity order, String newStatus) {
        String email = order.getBillingEmail() != null ? order.getBillingEmail().trim() : "";
        String phone = order.getBillingPhone() != null ? order.getBillingPhone().trim() : "";
        String name = (order.getBillingName() != null && !order.getBillingName().trim().isEmpty())
                ? order.getBillingName().trim()
                : "Quý khách";
        String body = buildGuestMessageBody(order.getId(), newStatus);

        if (!email.isEmpty()) {
            Log.i(TAG, "[MOCK EMAIL -> " + email + "] To: " + name + " <" + email + "> | Subject: Cập nhật đơn hàng #" + order.getId() + " | Body: " + body);
        } else if (!phone.isEmpty()) {
            Log.i(TAG, "[MOCK SMS -> " + phone + "] To: " + name + " | Body: " + body);
        } else {
            Log.w(TAG, "Guest order " + order.getId() + " has no email or phone on file; cannot route notification.");
        }
    }

    private static String buildGuestMessageBody(String orderId, String status) {
        return resolveNotificationContent(orderId, status).content;
    }

    private static String buildMessageBody(String orderId, String statusPhrase) {
        return "Rootie: Đơn hàng #" + orderId + " của bạn " + statusPhrase + ". Cảm ơn bạn đã mua sắm tại Rootie Vietnam.";
    }

    private static String currentEventTime() {
        return new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault()).format(new Date());
    }

    private static String resolveOrderPlacementTime(OrderEntity order) {
        if (order != null) {
            String fromFields = NotificationDateHelper.toStorageTimeFromOrder(
                    order.getOrderDate(),
                    order.getOrderTime()
            );
            if (fromFields != null) {
                return fromFields;
            }
            if (order.getCreatedAt() > 0L) {
                return new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
                        .format(new Date(order.getCreatedAt()));
            }
        }
        return currentEventTime();
    }

    private static final class NotificationContent {
        private final String title;
        private final String content;

        private NotificationContent(String title, String content) {
            this.title = title;
            this.content = content;
        }
    }
}

File: app/src/main/java/com/veganbeauty/app/data/repository/OrderStatusNotifier.kt
package com.veganbeauty.app.data.repository

import android.util.Log
import com.veganbeauty.app.data.local.dao.OrderDao
import com.veganbeauty.app.data.local.entities.OrderEntity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * Mock Order Status Update handler.
 *
 * Whenever the status of an order changes, this helper is invoked. It is
 * deliberately a stand-in for a real backend "order_status_updated" webhook:
 *
 *  - If the order has a [OrderEntity.userId] of `null` (i.e. a guest checkout):
 *      - If [OrderEntity.billingEmail] is present, a simulated **Email** message
 *        is printed to the backend (Logcat) console.
 *      - Otherwise, a simulated **SMS** message is printed.
 *  - If the order belongs to a registered user, no outbound notification is
 *    sent from here — that responsibility is delegated to the user-facing
 *    notification preferences (see [com.veganbeauty.app.data.local.ProfileSession]).
 *
 * No real SMS gateway or email provider is contacted. All output is
 * written to `Logcat` under the [TAG] tag so the team can verify the
 * notification routing from the [Verification Plan] test cases.
 */
object OrderStatusNotifier {

    private const val TAG = "OrderStatusNotifier"

    /** Fire-and-forget status update. Safe to call from any coroutine context. */
    fun updateStatus(
        orderDao: OrderDao,
        orderId: String,
        newStatus: String,
        scope: CoroutineScope = CoroutineScope(Dispatchers.IO)
    ) {
        scope.launch {
            try {
                val updated = orderDao.updateOrderStatus(orderId, newStatus)
                if (updated <= 0) {
                    Log.w(TAG, "updateStatus: no order found for id=$orderId")
                    return@launch
                }
                simulateNotificationForOrder(orderDao, orderId, newStatus)
            } catch (e: Exception) {
                Log.e(TAG, "updateStatus failed for orderId=$orderId", e)
            }
        }
    }

    /**
     * Dispatch the mock SMS/Email notification only. The caller is expected
     * to have already persisted the new status on the order. Useful when
     * the status update and the notification dispatch live in the same
     * suspend transaction (e.g. inside
     * [com.veganbeauty.app.data.repository.OrderRepository.updateOrderStatus]).
     */
    suspend fun simulateOnly(orderDao: OrderDao, orderId: String, newStatus: String) {
        try {
            simulateNotificationForOrder(orderDao, orderId, newStatus)
        } catch (e: Exception) {
            Log.e(TAG, "simulateOnly failed for orderId=$orderId", e)
        }
    }

    private suspend fun simulateNotificationForOrder(
        orderDao: OrderDao,
        orderId: String,
        newStatus: String
    ) {
        val order = orderDao.getOrderById(orderId) ?: return
        simulateOutboundNotification(order, newStatus)
    }

    /**
     * Prints a mock SMS/Email notification to Logcat. In a real backend this
     * would POST to a transactional email/SMS provider; here we just log the
     * payload so the Android team can verify the routing decision.
     */
    private fun simulateOutboundNotification(order: OrderEntity, newStatus: String) {
        val isGuest = order.isGuest || order.userId.isNullOrBlank()
        if (!isGuest) {
            Log.i(TAG, "Order ${order.id} → $newStatus (user=${order.userId}, " +
                "in-app notification dispatched via user preferences)")
            return
        }
        val email = order.billingEmail?.trim().orEmpty()
        val phone = order.billingPhone?.trim().orEmpty()
        val name = order.billingName?.trim().orEmpty().ifEmpty { "Quý khách" }
        val body = buildMessageBody(order.id, newStatus)

        if (email.isNotEmpty()) {
            // Email routing
            Log.i(
                TAG,
                "[MOCK EMAIL → $email] To: $name <$email> | Subject: " +
                    "Cập nhật đơn hàng #${order.id} | Body: $body"
            )
        } else if (phone.isNotEmpty()) {
            // SMS routing
            Log.i(
                TAG,
                "[MOCK SMS → $phone] To: $name | Body: $body"
            )
        } else {
            Log.w(
                TAG,
                "Guest order ${order.id} has no email or phone on file; " +
                    "cannot route notification."
            )
        }
    }

    private fun buildMessageBody(orderId: String, status: String): String {
        val friendlyStatus = when (status) {
            "Đang xử lý", "Đã xác nhận" -> "đã được xác nhận và đang được chuẩn bị"
            "Đang giao hàng", "Đang vận chuyển" -> "đang được giao đến bạn"
            "Đã giao hàng", "Hoàn tất" -> "đã được giao thành công"
            "Đã hủy" -> "đã được hủy theo yêu cầu"
            else -> "đã được cập nhật trạng thái: $status"
        }
        return "Rootie: Đơn hàng #$orderId của bạn $friendlyStatus. Cảm ơn bạn đã mua sắm tại Rootie Vietnam."
    }
}
